test_provider.py

```diff
@@ -0,0 +1,454 @@
+
+    def parse_or_group(self, json: dict) -> OrGroup:
+        
+        # Find nested OR groups/AND groups
+        # TODO: is this correct if there are not nested groups in OCS natively
+
+        observations = [self.parse_observation(json[key], key) for key in json.keys() if key.startswith('OBSERVATION_BASIC')]
+
+        number_to_observe = len(observations)
+        delay_max, delay_min = 0, 0 # TODO: What are these?
+        or_group = OrGroup(json['key'], json['name'], number_to_observe, delay_min, delay_max, observations)
+        return or_group
+
+    @staticmethod
+    def parse_guide_star(json: dict) -> Resource:
+
+        # TODO: Maybe a GuideStart class should be in place to handle this better
+        res = Resource(json['name'], json['tag'])
+        return res
+    
+    def parse_root_group(self, json: dict) -> AndGroup:
+        # Find nested OR groups/AND groups
+        groups = [self.parse_and_group(json[key]) for key in json.keys() if key.startswith(self._GroupsKeys.KEY.value)]
+        if any(key.startswith(self._GroupsKeys.ORGANIZATIONAL_FOLDER.value) for key in json.keys()):
+            for key in json.keys():
+                if key.startswith(self._GroupsKeys.ORGANIZATIONAL_FOLDER.value):
+                    groups.append(self.parse_or_group(json[key]))
+        num_to_observe = len(groups)
+        root_group = AndGroup(None, None, num_to_observe, 0, 0, groups, AndOption.ANYORDER)
+        return root_group
+
+    def parse_and_group(self, json: dict) -> AndGroup:
+        observations = [self.parse_observation(json[key], key) for key in json.keys() if key.startswith(self._ObsKeys.KEY.value)]
+
+        number_to_observe = len(observations)
+        delay_max, delay_min = 0, 0 # TODO: What are these?
+        return AndGroup(json['key'], json['name'], number_to_observe, delay_min, delay_max, observations, AndOption.ANYORDER)
+
+    def parse_program(self, json: dict) -> Program:
+        
+        too_type = TooType(json[self._ProgramKeys.ToO.value]) if json[self._ProgramKeys.ToO.value] != 'None' else None
+        ta = self.parse_time_allocation(json)
+        root_group = self.parse_root_group(json)
+        id = json[self._ProgramKeys.ID.value]
+        program_type =  self.program_types[id.split('-')[2]]
+
+        notes = [json[key] for key in json.keys() if key.startswith(self._ProgramKeys.NOTES.value)]
+
+        start, end = self.get_program_dates(program_type, id, notes)
+        
+        return Program(id,
+                       json[self._ProgramKeys.INTERNAL_ID.value],
+                       Band(int(json[self._ProgramKeys.BAND.value])),
+                       bool(json[self._ProgramKeys.THESIS.value]),
+                       ProgramMode[json[self._ProgramKeys.MODE.value].upper()],
+                       program_type,
+                       start,
+                       end,
+                       ta,
+                       root_group,
+                       too_type)
```

Comment:
This is good, and how I'd like to see the mini-model classes instantiated and returned.